[quant][gpu][core] Implemented quantized add operator using cudnn [reland PR74463]

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74463

Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35218224](https://our.internmc.facebook.com/intern/diff/D35218224)

dzdang committed:
[quant][gpu][core] Implemented quantized added operator in cudnn

Summary:
This PR implements the quantized add operator using cudnn operations.
A test case was added ....

Test Plan:
TBA

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized added operator in cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
A test case was added ....

Test Plan:
TBA

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized added operator in cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
A test case was added ....

Test Plan:
TBA

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized added operator in cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
A test case was added ....

Test Plan:
TBA

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized added operator in cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized added operator in cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -0,0 +1,210 @@
+    while (new_sizes.size() < 3) {
+      new_sizes.emplace_back(1);
+    }
+    qa = at::native::view(qa, new_sizes);
```
nit: I think we should be using at::view, also we should be able to do `qa.view(....)` as well I think

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator in cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator in cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35009111).

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -0,0 +1,217 @@
+
+namespace at {
+namespace native {
+namespace {
```
why do we have two anonymous namespaces? should we merge them?

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -829,6 +829,44 @@ def test_qadd_relu_same_qparams(self):
+    """Tests the correctness of the cudnn add and add_relu op
+    (Similar to test_qadd_relu_different_qparams, will probably merge in the future)"""
+    @unittest.skipIf(not TEST_CUDNN, "cudnn is not enabled.")
+    @unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
```
what is the effort to make this a flag accessible in pytorch, like TEST_CUDNN? I feel ideally we don't want to copy paste the code for every operator

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35009111).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -0,0 +1,222 @@
+  auto orig_sizes = qa.sizes().vec();
+  if (qa.dim() < 3) {
+    std::vector<int64_t> new_sizes{orig_sizes};
+    while (new_sizes.size() < 3) {
+      new_sizes.emplace_back(1);
+    }
+    // I think cudnn expects leading dimensions to be the dummy dimensions, so we need to swap
+    std::swap(new_sizes.front(), new_sizes.back());
```
can we just use insert to do this?

jerryzh168 replied:
oh, I think a cleaner way is to initialize a vector with the expected length and assign the last dim to orig_sizes

jerryzh168 reviewed (approved):
thanks for the comments for overall and individual ops, this makes it much easier to follow

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -0,0 +1,222 @@
+// Relu op: relu_op
+template <bool kReluFused = false>
+Tensor add(Tensor qa, Tensor qb, double output_scale, int64_t output_zero_point) {
+  if (qa.numel() == 0) {
+    return Tensor{};
+  }
```
should we check both qa and qb? and when does this happen?

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35009111).

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35009111).

jjsjann123 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -0,0 +1,224 @@
+    std::vector<int64_t> new_sizes(3, 1);
+    // cudnn expects leading dimensions to be the dummy dimensions
+    new_sizes.back() = qa.sizes().back();
+    if (qa.ndim() == 2) {
```
This line should be `qa.dim()` instead, similar to previous line 67.

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn [reland PR74463]"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn [reland PR74463]"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```
a

Differential Revision: [D35009111](https://our.internmc.facebook.com/intern/diff/D35009111)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][gpu][core] Implemented quantized add operator using cudnn [reland PR74463]"


Summary:
This PR implements the quantized add operator using cudnn operations.
Also added a corresponding test function in test_quantized_op.py. Ideally,
we should merge this function with the cpu variant, but for now, we will
keep it separate until cudnn v8 is in the default build. Other factors also
complicate the merge as cudnn quantized add is currently only supported for
int8 symmetrically quantized tensors.

Test Plan:
In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedOps.test_qadd_relu_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35218224).
